Sign in with Apple post, version 1 (#106)
xmartlabs/blog: _posts/2020-05-04-Why-Sign-in-with-Apple-and-integration-guide.md

---
layout: post
title: "Sign in with Apple! Is it necessary? at which cost?"
excerpt: "In this blogpost we talk about the benefits of providing Sign in with Apple in your app. We also provide a step by step guid on how to integrate it"
date: 2020-05-04 10:00:00
author: Cecilia Pirotto
tags: [Xmartlabs, iOS, Apple, Swift, Sign in with Apple]
category: development
author_id: ceci
featured_image: /images/apple-sign-in/signInWithApple.jpg
show: true
crosspost_to_medium: false

---

In this post, we'll debate about the benefits of providing *Sign in with Apple* in your app so you can decide if it worth integrating it or not. In the second part of this post we'll provide a step by step *Sign in with Apple* integration guide and talk about all the issues we run into and had to overcome.

### In case you don't know What's Sign in with Apple yet...

> It's a new third party login service provided by Apple where users can sign in into your app using their Apple ID.

Reviewer: Maybe authentication service instead of login service? Login service seems to be something that manage a user session, in this case Sign in with apple only provides us a way to validate that someone is who they say they are.

### What does *Sign in with Apple* put on the table?

It provides a one tap frictionless login and authentication system to your app which means more people will login into your app and also a faster growth in the number of app users especially in Apple device owners who only need to check their identity through Touch Id or Face Id. By using *Sign in with Apple*, users don't need to remember app credentials, apps don't need to provide a password reset and identity and validation workflow in the app, neither provide a specific register and login form.

*Sign in with Apple* is FIDO U2F standard complaint, which means security aspects are met and we don't need to care about it. Apple adds two-factor authentication support by default, providing an extra layer of security.

Sometimes an app user would prefer *Sign in with Apple* over other alternatives because it has the ability to hide its real email, this still allows the app to reach the user real email through Apple servers. Apple provides a user's private email that is only reachable from the app registered email domains, so the user email doesn't have value outside app servers and can't be sold.

> Even though *Sign in with Apple* is multiplatform which means we can make it work (in addition to the platforms provided by Apple) on the web, Android devices, and Windows apps. The user still needs to have an Apple device to complete the two-factor authentication, upon Apple Id login the user receives a 2FA code from apple in their device. So if your app is available for not Apple devices owners just allowing *Sign in with Apple* is not an option.

Reviewer: ```suggestion
Even though *Sign in with Apple* is multiplatform which means we can make it work on the web, Android devices, Windows apps and platforms provided by Apple. The user still needs to have an Apple device to complete the two-factor authentication, upon Apple Id login the user receives a 2FA code from apple in their device. So if your app is available for not Apple devices owners just allowing *Sign in with Apple* is not an option.
```

Also not sure what do you mean with `upon Apple Id login`

<img width="100%" src="/images/apple-sign-in/registerNewKey.png" />

Make sure to select the correct Primary App ID and save the configuration key.

<img width="100%" src="/images/apple-sign-in/configurekey.png" />

Now, we have the environment setup done, let's get into the code.

### Add button

Although you can use your custom button, we strongly recommend using those provided by Apple which offers multiple benefits shown [here](https://developer.apple.com/design/human-interface-guidelines/sign-in-with-apple/overview/buttons/).

<div style="text-align: center"><img width="30%" src="/images/apple-sign-in/signInButton.png" /></div>
<p></p>
It's necessary to import `AuthenticationServices` framework which provides `ASAuthorizationAppleIDButton` button class. Adding this button is very simple, you just need to create a button instance and set up its `touchUpInside` handler. Remember to add the button to the view hierarchy.

```swift
  let button = ASAuthorizationAppleIDButton()
  button.addTarget(self, action: #selector(handleAuthorizationAppleIDButtonPress), for: .touchUpInside)
  self.loginProviderStackView.addArrangedSubview(button)
```

### Handle button press

> After we add the button, we need to implement the button tap handler. We'll will request *Sign in with Apple* permissions using `ASAuthorizationAppleIDProvider` class and we will use `ASAuthorizationController` to to handle authentication interface flow.  

Reviewer: ```suggestion
After we add the button, we need to implement the button tap handler. We'll request *Sign in with Apple* permissions using `ASAuthorizationAppleIDProvider` class and we'll use `ASAuthorizationController` to handle the authentication interface flow.  
```

      let request = ASAuthorizationAppleIDProvider().createRequest() //1
      request.requestedScopes = [.fullName, .email] //2

      let authorizationController = ASAuthorizationController(authorizationRequests: [request]) //3
      authorizationController.delegate = self
      authorizationController.presentationContextProvider = self //4
      authorizationController.performRequests() //5
  }
```
1. Create a request (`ASAuthorizationAppleIDRequest`). To create it we need an `ASAuthorizationAppleIDProvider` instance.
2. Define the request scope, which is the data we want to receive from the user (in this case email and full name).
3. Create a controller that manages authorization requests created by a provider.
4. Provides a display context to present the authorization interface to the user.
5. Perform the request and opens the authentication dialog shown below.

<div style="text-align: center"><img width="60%" src="/images/apple-sign-in/AppleSignIn.gif" /></div>
<p></p>
As we can see int he code snippet above, it's required to conform to `ASAuthorizationControllerDelegate` and `ASAuthorizationControllerPresentationContextProviding` protocols.
<p></p>

> we conform to `ASAuthorizationControllerPresentationContextProviding` to indicate the window that will contain "Sign in with Apple" SDK dialogs, as shown in the code below.

Reviewer: ```suggestion
We conform to `ASAuthorizationControllerPresentationContextProviding` to indicate the window that will contain "Sign in with Apple" SDK dialogs, as shown in the code below.
```

```swift
  // ASAuthorizationControllerPresentationContextProviding
  func presentationAnchor(for controller: ASAuthorizationController) -> ASPresentationAnchor {
      return self.view.window
  }
```

### Which user authentication data does we get from Apple?

We'll receive an `ASAuthorizationAppleIDCredential` instance, here are the principal information:

* **User ID**: The unique user ID
* **Full name**: User could edit his full name before sharing it with your app

> * **Email**: A user's email address, which could either be the real Apple Id user email or an obscured one

Reviewer: ```suggestion
* **Email**: A user's email address, which could either be the real user email or an obscured one
```

> * **Authorization Code & Identity Token**: These are encrypted data your client app should send to the backend so it  can validate the credentials.

Reviewer: ```suggestion
* **Authorization Code & Identity Token**: These are encrypted data your client app should send to the backend so it  can validate the information's authenticity.
```

It may sound confusing to say that the app's backend validates the user's credentials, one may think that the email/username and password for apple are also being sent to the backend.

> *Full name* and *Email* are optional and only available for new users, which means, users that are login into the app for the first time.

Reviewer: ```suggestion
*Full name* and *Email* are optional and only available for new users, which means, users that are logging into the app for the first time.
```
